### PR TITLE
[FLINK-4364] [runtime] [FLIP-6] Implement TaskManager side of heartbeat from JobManager

### PR DESCRIPTION
**Pull request by zhijiangW**
1. `TaskManagerRunner` creates `HeartbeatManagerImpl` components when constructs the **TaskManager**.
2. `TaskManager` begins to monitor the `JobManager` when registration at new job leader successful.
3. Currently the payload will be null for interaction between `TaskManager` and `JobManager`, it can be expanded if needed later. 

@tillrohrmann  for review please!

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterGateway.java [hunk 1]
```diff
@@ -218,4 +218,12 @@ void failSlot(final ResourceID taskManagerId,
...
 			final UUID leaderId,
 			@RpcTimeout final Time timeout);
+
+	/**
+	 * Send the heartbeat to job manager from task manager
+	 *
+	 * @param resourceID unique id of the task manager
+	 * @param payload the payload information from the task manager
+	 */
+	void heartbeatFromTaskManager(final ResourceID resourceID, final Object payload);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterGateway.java]: The type of the payload should not be `Object`

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java [hunk 2]
```diff
@@ -126,6 +129,9 @@
 	/** The metric registry in the task manager */
 	private final MetricRegistry metricRegistry;
 
+	/** The heartbeat manager for job manager and resource manager in the task manager */
+	private final HeartbeatManagerImpl heartbeatManager;
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: Raw usage of `HeartbeatManagerImpl`

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java [hunk 3]
```diff
@@ -1058,6 +1092,30 @@ public void handleError(Throwable throwable) {
 		}
 	}
 
+	/**
+	 * The heartbeat listener for JobManager and ResourceManager, they can be distinguished by ResourceID
+	 * and trigger different processes.
+	 */
+	private final class JMRMHeartbeatListener implements HeartbeatListener {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: Why `JMRM` and not `JMTM`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java [hunk 4]
```diff
@@ -1058,6 +1092,30 @@ public void handleError(Throwable throwable) {
...
+
+		JMRMHeartbeatListener() {
+		}
+
+		@Override
+		public void notifyHeartbeatTimeout(final ResourceID resourceID) {
+			log.info("Notify heartbeat timeout for resourceID {}", resourceID);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: A heartbeat timeout should trigger some kind of action. E.g. unregistering the JM.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java [hunk 5]
```diff
@@ -112,6 +114,13 @@ public TaskManagerRunner(
 		// Initialize the TM metrics
 		TaskExecutorMetricsInitializer.instantiateStatusMetrics(taskManagerMetricGroup, taskManagerServices.getNetworkEnvironment());
 
+		HeartbeatManagerImpl heartbeatManager = new HeartbeatManagerImpl(
+				taskManagerConfiguration.getTimeout().toMilliseconds(),
+				resourceID,
+				executor,
+				Executors.newSingleThreadScheduledExecutor(),
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: Maybe we should introduce a `ScheduledExecutorService` which can be shared across components and which is given to the respective `Runner`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java [hunk 6]
```diff
@@ -741,6 +763,18 @@ private void establishJobManagerConnection(JobID jobId, JobMasterGateway jobMast
...
 
+		heartbeatManager.monitorTarget(registrationSuccess.getResourceID(), new HeartbeatTarget() {
+			@Override
+			public void sendHeartbeat(ResourceID resourceID, Object payload) {
+				jobMasterGateway.heartbeatFromTaskManager(resourceID, payload);
+			}
+
+			@Override
+			public void requestHeartbeat(ResourceID resourceID, Object payload) {
+				throw new UnsupportedOperationException("Should never call requestHeartbeat in task manager.");
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: A heartbeat request is totally fine. This should not throw an `UnsupportedOperationException` but trigger a heartbeat being sent back to `resourceID`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterGateway.java (hunk 1 above)
Comment (wangzhijiang999) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterGateway.java]: Currently we have not considered which specific payload information should be attached with the heartbeat, so use the object to work around.  It should be changed to specific type if confirmation. Do you think which payload should be attached necessary currently?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java (hunk 3 above)
Comment (wangzhijiang999) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: From TaskExecutor side, it will monitor the JM and RM that initiate the heartbeat, so JMRMHeartbeatListener indicates the heartbeat timeout with JM or RM, the current PR just shows the interaction with JM part. I am supposed to submit the RM part in another PR, maybe submit both sides together will help to understand easily.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java (hunk 4 above)
Comment (wangzhijiang999) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: Yes, it actually should trigger some actions with timeout. Currently I did not submit this part because I think it is related with failure detection logic  and supposed to submit in another PR. To make the heartbeat mechanism complete, I will add this part in the following modifications.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java (hunk 5 above)
Comment (wangzhijiang999) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: I agree with that, and initially I want to reuse the executor in **RPCService** but it needs some modifications. I will introduce it in Runner in order to share among components.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java (hunk 6 above)
Comment (wangzhijiang999) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: My previous understanding is that the heartbeat is only requested from RM and JM to TM, and the TM will only response the heartbeat. Do you mean the heartbeat can both request from both sides? If to do so, the TM also needs to schedule a heartbeat request at interval time.

**Comment by wangzhijiang999** [comment]
@tillrohrmann . Thank you for detail review and comments!

This PR just submit the heartbeat logic in TM side, because there is already a jira of JM heartbeat side.

For my implementation, the JM initiates the heartbeat with **HeartbeatManagerSenderImpl** and the TM responses the heartbeat with **HeartbeatManagerImpl**. So the heartbeat process is one-way.

I think it is better to submit the JM heartbeat logic in this PR in order to understand easily. I will modify this PR soon, and for testing there already exists the UT for basic heartbeat logic. Do you mean to add some ITCases?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java (hunk 2 above)
Comment (wangzhijiang999) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: Yes, the current **HeartbeatManagerImpl** is enough to use, so I did not wrapper it.

Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: No I meant that you did not specify the generic type parameters.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterGateway.java (hunk 1 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterGateway.java]: If it's `Void` at the moment, why not leaving this parameter out for the moment?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java (hunk 6 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: Sure, but a JM and RM should also be able to send their own heartbeat without a request for an answer to the TM, right? What I'm trying to say is that the TM should not constrain unnecessarily the way you can interact with it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java (hunk 3 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: Hmm it could make sense to have a version specialized for the JM and RM because you will trigger different operations. And then the TM doesn't have to make the distinction.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java (hunk 4 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: I think it would be enough to create the respective handler method in the `TaskExecutor`. The actual handling can then be implemented in a separate PR. But that way, we can easily test the behaviour by checking whether this method was called in case of a timeout.

**Comment by tillrohrmann** [comment]
Thanks @wangzhijiang999. Ping me once you've updated the PR. I think it would make sense to test that the respective steps are taken if the JM detects that a TM died and vice versa. For that you basically only need to start a JM and a TM and let them connect.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java (hunk 6 above)
Comment (wangzhijiang999) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java]: If I understand correctly the TM can request the heartbeat to itself and the JM also can send the heartbeat to itself? 
If it is true, the **requestHeartbeat** will never be called from current logic in TM side. Or I can just remove the **UnsupportedOperationException** to not limitation and leave this method empty currently?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java (hunk 5 above)
Comment (wangzhijiang999) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: Do you have further suggestions where to introduce the **ScheduledExecutorService** seems suitable?

**Comment by wangzhijiang999** [comment]
@tillrohrmann , I am preparing the testing code and can submit the updates this week. Thank you for continuous help!

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java (hunk 5 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: Maybe we could add to the `RpcService` that it provides a `ScheduledExecutorService`. If we see that the load on this executor is too much, then we can introduce a dedicated service. What do you think?

Comment (wangzhijiang999) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: Yes, I totally agree with that and the advantages can avoid bringing extra thread pool and run all the PRC messages in the uniform executor if **RPCService** can handle well.

**Comment by wangzhijiang999** [comment]
@tillrohrmann , I have submitted the updates that may cover your suggestions.

There are still two issues that I am not confirmed.

First, for heartbeat interval and timeout default values in **ConfigConstants**, they are not invalid currently and you can modify it with your professional experience.

Second, the introduction of **ScheduledExecutorService** in **RPCService**, my initial idea is trying to use the existing scheduler in **RPCService**, but it can not be got from **AkkaRPCService** implementation. Another way is to replace the current **ScheduledExecutorService** parameter with **RPCService** in construction of **HeartbeatManagerSenderImpl**, and the **RPCService** can also schedule the heartbeat request. But the return value of **scheduleRunnable** method in **RPCService** is conflict with that in **HeartbeatManagerSenderImpl**. So I just bring another single thread pool in **RPCService** for use currently. Maybe the number of threads in pool can refer to number of cpu cores.

Maybe there are still something to be polished, and I am willing for further modifications by your comments.
BTW, the heartbeat interaction between TM and RM will be submitted in another PR after this confirmation because of some common points.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+			// before heartbeat timeout
+			assertTrue(heartbeatTargets.containsKey(tmResourceId));
+			assertTrue(registeredTMsInJM.containsKey(tmResourceId));
+			assertTrue(registeredTMsInSlotPool.contains(tmResourceId));
+
+			// trigger heartbeat timeout in job manager side, because the task manager will not response the heartbeat
+			Thread.sleep(heartbeatTimeout);
+
+			// after heartbeat timeout
+			verify(taskExecutorGateway, atLeast(1)).heartbeatFromJobManager(eq(jmResourceId));
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java]: In order to wait for the timeout being triggered you could extend the `HeartbeatManagerImpl` and introduce a wait mechanism in `unmonitorTarget` method. This should be called by the `JobMaster` to unregister the timed out `TaskExecutor`.

**Comment by zhijiangW** [comment]
Hi, @tillrohrmann , thanks for so detail reviews! I got all the points and may submit the modifications tomorrow, then I will ping you!

**Comment by zhijiangW** [comment]
@tillrohrmann , I have submitted the modifications that covered all your suggestions. 

Only one issue I am not confirmed is in order to reuse the **ScheduledExecutor** in **RpcService**, I modified the type of **ScheduledExecutorService** in **HeartbeatManagerImpl**, otherwise we need another wrapper like **ScheduledExecutorServiceAdapter**.

Wish your further advices and thanks in advance!

**Comment by tillrohrmann** [comment]
Hi @zhijiangW, just wanted to check on you what the state of this PR is?

**Comment by zhijiangW** [comment]
Hi @tillrohrmann , I have submitted the modifications as you suggest one week ago. Have you received it and any other issues?

**Comment by tillrohrmann** [comment]
Ah perfect. Sorry my bad. Will take a look and try to merge it this week :-) Thanks a lot for your work!

**Comment by zhijiangW** [comment]
@tillrohrmann , just remind to review my modifications for your free time, because this would block my next pull request of heartbeat between `TaskManager` and `ResourceManager`. Thank you!

**Comment by tillrohrmann** [comment]
Hi @zhijiangW, I'm currently merging the PR. Please also open the other PRs for the heartbeats between the RM <--> JM and RM <--> TM. This makes it easier for me to quickly merge all of them in one fell swoop.

**Comment by zhijiangW** [comment]
@tillrohrmann , yeah, I will submit the following PRs soon after this one is merged into master, because there are some common codes base on this PR.  Thank you for merging!
